core/src/main/java/com/orientechnologies/orient/core/delta/ODocumentDeltaSerializerV1.java

```diff
@@ -0,0 +1,583 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.orientechnologies.orient.core.delta;
+
+import com.orientechnologies.common.collection.OMultiValue;
+import com.orientechnologies.common.serialization.types.OByteSerializer;
+import com.orientechnologies.common.serialization.types.ODecimalSerializer;
+import com.orientechnologies.common.serialization.types.OIntegerSerializer;
+import com.orientechnologies.common.serialization.types.OLongSerializer;
+import com.orientechnologies.orient.core.db.record.*;
+import com.orientechnologies.orient.core.db.record.ridbag.ORidBag;
+import com.orientechnologies.orient.core.exception.OSerializationException;
+import com.orientechnologies.orient.core.exception.OValidationException;
+import com.orientechnologies.orient.core.metadata.schema.OClass;
+import com.orientechnologies.orient.core.metadata.schema.OGlobalProperty;
+import com.orientechnologies.orient.core.metadata.schema.OProperty;
+import com.orientechnologies.orient.core.metadata.schema.OType;
+import com.orientechnologies.orient.core.record.impl.ODocument;
+import com.orientechnologies.orient.core.record.impl.ODocumentEntry;
+import com.orientechnologies.orient.core.record.impl.ODocumentInternal;
+import com.orientechnologies.orient.core.serialization.ODocumentSerializable;
+import com.orientechnologies.orient.core.serialization.OSerializableStream;
+import com.orientechnologies.orient.core.serialization.serializer.record.binary.BytesContainer;
+import com.orientechnologies.orient.core.serialization.serializer.record.binary.HelperClasses;
+import com.orientechnologies.orient.core.serialization.serializer.record.binary.OSerializableWrapper;
+import com.orientechnologies.orient.core.serialization.serializer.record.binary.OVarIntSerializer;
+import com.orientechnologies.orient.core.util.ODateHelper;
+
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.util.*;
+
+import static com.orientechnologies.orient.core.serialization.serializer.record.binary.HelperClasses.*;
+
+/**
+ * @author mdjurovi
+ */
+public class ODocumentDeltaSerializerV1 extends ODocumentDeltaSerializer {
+
+  protected ODocumentDeltaSerializerV1() {
+
+  }
+
+  @Override
+  public byte[] toStream(ODocumentDelta delta) {
+    BytesContainer bytes = new BytesContainer();
+    serialize(delta, bytes);
+    return bytes.fitBytes();
+  }
+
+  private void serialize(ODocumentDelta deltaDoc, BytesContainer bytes) {
+    for (Map.Entry<String, Object> entry : deltaDoc.fields.entrySet()) {
+      String fieldName = entry.getKey();
+      Object fieldValue = entry.getValue();
+
+      //serialize field name
+      serializeValue(bytes, fieldName, OType.STRING);
+
+      //serialize field type
+      OType type = OType.getTypeByClass(fieldValue != null ? fieldValue.getClass() : null);
```

Comment:
this is semantically correct, anyway have the type directly in ODocumentDelta sourced directly from the ODocument may be more efficient.

Reply:
I've done this optimization and right now delta document field carries type of field fetched from ODocument (when it is possible)